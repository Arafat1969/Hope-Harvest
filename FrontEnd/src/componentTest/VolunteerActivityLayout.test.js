import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import VolunteerActivityLayout from '../components/VolunteerActivityLayout';

// Create a mock for useNavigate before importing the component
const mockNavigate = jest.fn();

// Mock react-router-dom
jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  MemoryRouter: ({ children }) => <div data-testid="memory-router">{children}</div>
}));

describe('VolunteerActivityLayout', () => {
  beforeEach(() => {
    mockNavigate.mockClear();
  });

  const renderComponent = () => {
    return render(<VolunteerActivityLayout />);
  };

  test('renders hero title and statistics', () => {
    renderComponent();

    expect(screen.getByText(/volunteer with hope harvest/i)).toBeInTheDocument();
    
    // Use getAllByText for duplicate elements and check specific ones
    const statisticElements = screen.getAllByText(/150K\+/);
    expect(statisticElements[0]).toBeInTheDocument();
    
    const hundredKElements = screen.getAllByText(/100K\+/);
    expect(hundredKElements[0]).toBeInTheDocument();
    
    // For elements that appear multiple times, use more specific selectors
    expect(screen.getByText('Hours Served')).toBeInTheDocument();
    expect(screen.getByText('Lives Touched')).toBeInTheDocument();
    expect(screen.getByText('Students Helped')).toBeInTheDocument();
    expect(screen.getByText('Patients Served')).toBeInTheDocument();
  });

  test('renders all step cards using role selectors', () => {
    renderComponent();

    // Use getAllByText to handle duplicates and select the step cards specifically
    const aimElements = screen.getAllByText('Aim');
    expect(aimElements.length).toBeGreaterThanOrEqual(1);
    
    const activityElements = screen.getAllByText('Activity');
    expect(activityElements.length).toBeGreaterThanOrEqual(1);
    
    const requirementsElements = screen.getAllByText('Requirements');
    expect(requirementsElements.length).toBeGreaterThanOrEqual(1);
    
    const achievementsElements = screen.getAllByText('Achievements');
    expect(achievementsElements.length).toBeGreaterThanOrEqual(1);
  });

  test('displays default step content (Aim)', () => {
    renderComponent();

    // Look for the content heading specifically (h3 level)
    const aimHeadings = screen.getAllByRole('heading', { level: 3 });
    const aimContentHeading = aimHeadings.find(heading => heading.textContent === 'Aim');
    expect(aimContentHeading).toBeInTheDocument();
    
    expect(screen.getByText(/The volunteer program at Hope Harvest aims to create/i)).toBeInTheDocument();
  });

  test('clicking a step card updates the step content', () => {
    renderComponent();

    // Find Activity in step cards (h6 elements) and click the first one
    const activityElements = screen.getAllByText('Activity');
    const activityStepCard = activityElements.find(el => el.tagName === 'H6');
    fireEvent.click(activityStepCard);

    // Check if Activity content heading (h3) is displayed
    const activityHeadings = screen.getAllByRole('heading', { level: 3 });
    const activityContentHeading = activityHeadings.find(heading => heading.textContent === 'Activity');
    expect(activityContentHeading).toBeInTheDocument();
    
    expect(screen.getByText(/Volunteers at Hope Harvest engage in a wide range of activities/i)).toBeInTheDocument();
  });

  test('clicking navigation dots updates active step', () => {
    renderComponent();

    // Find step navigation elements by their step text
    const stepElements = screen.getAllByText(/Step \d/);
    expect(stepElements.length).toBe(4);
    
    // Click the second step (Activity)
    const step2Element = screen.getByText('Step 2');
    const activityCard = step2Element.parentElement;
    fireEvent.click(activityCard);

    // Check if Activity content is displayed
    const activityHeadings = screen.getAllByRole('heading', { level: 3 });
    const activityContentHeading = activityHeadings.find(heading => heading.textContent === 'Activity');
    expect(activityContentHeading).toBeInTheDocument();
  });

  test('step cards have proper styling and hover effects', () => {
    renderComponent();

    // Find the step card by looking for Step 1 text and its parent
    const step1Element = screen.getByText('Step 1');
    const aimCard = step1Element.parentElement;
    
    // Test hover effect
    fireEvent.mouseEnter(aimCard);
    fireEvent.mouseLeave(aimCard);
    
    expect(aimCard).toBeInTheDocument();
  });

  test('progress bar updates based on active step', () => {
    renderComponent();

    // Find Requirements step card and click it
    const requirementsElements = screen.getAllByText('Requirements');
    const requirementsStepCard = requirementsElements.find(el => el.tagName === 'H6');
    fireEvent.click(requirementsStepCard);

    // Check if the step content changed
    const requirementsHeadings = screen.getAllByRole('heading', { level: 3 });
    const requirementsContentHeading = requirementsHeadings.find(heading => heading.textContent === 'Requirements');
    expect(requirementsContentHeading).toBeInTheDocument();
  });

  test('all statistics are displayed correctly', () => {
    renderComponent();

    expect(screen.getByText('Hours Served')).toBeInTheDocument();
    expect(screen.getByText('Lives Touched')).toBeInTheDocument();
    expect(screen.getByText('Students Helped')).toBeInTheDocument();
    expect(screen.getByText('Patients Served')).toBeInTheDocument();
  });

  test('clicking "Join as Volunteer" navigates to registration page', () => {
    renderComponent();

    const joinButton = screen.getByRole('button', { name: /Join as Volunteer/i });
    fireEvent.click(joinButton);

    expect(mockNavigate).toHaveBeenCalledWith('/volunteer-registration');
  });

  test('join button has proper styling and hover effects', () => {
    renderComponent();

    const joinButton = screen.getByRole('button', { name: /Join as Volunteer/i });
    
    // Test hover effects
    fireEvent.mouseEnter(joinButton);
    fireEvent.mouseLeave(joinButton);
    
    expect(joinButton).toBeInTheDocument();
  });

  test('component renders without crashing', () => {
    renderComponent();
    
    expect(screen.getByText('Volunteer with Hope Harvest')).toBeInTheDocument();
    expect(screen.getByText('Explore Volunteer Journey')).toBeInTheDocument();
  });

  test('all step content is accessible', () => {
    renderComponent();

    const steps = [
      { name: 'Aim', stepNumber: 'Step 1' },
      { name: 'Activity', stepNumber: 'Step 2' },
      { name: 'Requirements', stepNumber: 'Step 3' },
      { name: 'Achievements', stepNumber: 'Step 4' }
    ];
    
    steps.forEach((step) => {
      // Find and click the step card using step number
      const stepElement = screen.getByText(step.stepNumber);
      const stepCard = stepElement.parentElement;
      fireEvent.click(stepCard);
      
      // Check if the content heading appears
      const contentHeadings = screen.getAllByRole('heading', { level: 3 });
      const contentHeading = contentHeadings.find(heading => heading.textContent === step.name);
      expect(contentHeading).toBeInTheDocument();
    });
  });

  test('displays step numbers correctly', () => {
    renderComponent();

    // Check if step numbers are displayed (1-4)
    expect(screen.getByText('Step 1')).toBeInTheDocument();
    expect(screen.getByText('Step 2')).toBeInTheDocument();
    expect(screen.getByText('Step 3')).toBeInTheDocument();
    expect(screen.getByText('Step 4')).toBeInTheDocument();
  });

  test('statistics section has proper structure', () => {
    renderComponent();

    // Test that statistics are properly grouped
    const statisticsContainer = screen.getByText('150K+').parentElement;
    expect(within(statisticsContainer).getByText('Hours Served')).toBeInTheDocument();
  });

  test('step cards section has proper structure', () => {
    renderComponent();

    // Test that step cards are properly structured
    expect(screen.getByText('Explore Volunteer Journey')).toBeInTheDocument();
    
    // Verify all step numbers exist
    for (let i = 1; i <= 4; i++) {
      expect(screen.getByText(`Step ${i}`)).toBeInTheDocument();
    }
  });

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});