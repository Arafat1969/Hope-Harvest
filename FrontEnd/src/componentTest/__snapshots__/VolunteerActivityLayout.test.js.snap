// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VolunteerActivityLayout matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="min-height: 100vh; padding-top: 2rem; padding-bottom: 2rem;"
  >
    <div
      class="container"
    >
      <div
        style="border-radius: 25px; padding: 4rem 2rem; color: white; margin-bottom: 3rem; position: relative; overflow: hidden; text-align: center;"
      >
        <i
          class="fas fa-users"
          style="position: absolute; top: -50px; right: -50px; opacity: 0.1; font-size: 20rem; transform: rotate(-15deg);"
        />
        <h1
          class="display-3 fw-bold mb-4"
        >
          Volunteer with Hope Harvest
        </h1>
        <p
          class="lead mb-4"
          style="font-size: 1.3rem;"
        >
          Join our dedicated community of volunteers making a real difference across Bangladesh
        </p>
        <div
          style="display: grid; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr)); gap: 1rem; margin-bottom: 2rem;"
        >
          <div
            style="color: white; border-radius: 15px; padding: 1.5rem; text-align: center; box-shadow: 0 5px 20px rgba(0,0,0,0.1);"
          >
            <h3>
              150K+
            </h3>
            <p
              class="mb-0"
            >
              Hours Served
            </p>
          </div>
          <div
            style="color: white; border-radius: 15px; padding: 1.5rem; text-align: center; box-shadow: 0 5px 20px rgba(0,0,0,0.1);"
          >
            <h3>
              100K+
            </h3>
            <p
              class="mb-0"
            >
              Lives Touched
            </p>
          </div>
          <div
            style="color: white; border-radius: 15px; padding: 1.5rem; text-align: center; box-shadow: 0 5px 20px rgba(0,0,0,0.1);"
          >
            <h3>
              5K+
            </h3>
            <p
              class="mb-0"
            >
              Students Helped
            </p>
          </div>
          <div
            style="color: white; border-radius: 15px; padding: 1.5rem; text-align: center; box-shadow: 0 5px 20px rgba(0,0,0,0.1);"
          >
            <h3>
              35K+
            </h3>
            <p
              class="mb-0"
            >
              Patients Served
            </p>
          </div>
        </div>
      </div>
      <div
        class="row mb-4"
      >
        <div
          class="col-12"
        >
          <h4
            class="text-center mb-4"
            style="color: rgb(46, 125, 50); font-weight: 700;"
          >
            Explore Volunteer Journey
          </h4>
        </div>
        <div
          class="col-md-6 col-lg-3 mb-3"
        >
          <div
            style="color: white; border-radius: 15px; padding: 1.5rem; cursor: pointer; transition: all 0.3s ease; box-shadow: 0 5px 20px rgba(0,0,0,0.1); text-align: center;"
          >
            <i
              class="fas fa-bullseye"
              style="font-size: 2rem; margin-bottom: 1rem;"
            />
            <h6
              style="font-weight: 600; margin-bottom: 0.5rem;"
            >
              Aim
            </h6>
            <small>
              Step 1
            </small>
          </div>
        </div>
        <div
          class="col-md-6 col-lg-3 mb-3"
        >
          <div
            style="background: white; color: rgb(51, 51, 51); border-radius: 15px; padding: 1.5rem; cursor: pointer; transition: all 0.3s ease; box-shadow: 0 5px 20px rgba(0,0,0,0.1); border: 2px solid #e8f5e9; text-align: center;"
          >
            <i
              class="fas fa-hands-helping"
              style="font-size: 2rem; margin-bottom: 1rem;"
            />
            <h6
              style="font-weight: 600; margin-bottom: 0.5rem;"
            >
              Activity
            </h6>
            <small>
              Step 2
            </small>
          </div>
        </div>
        <div
          class="col-md-6 col-lg-3 mb-3"
        >
          <div
            style="background: white; color: rgb(51, 51, 51); border-radius: 15px; padding: 1.5rem; cursor: pointer; transition: all 0.3s ease; box-shadow: 0 5px 20px rgba(0,0,0,0.1); border: 2px solid #e8f5e9; text-align: center;"
          >
            <i
              class="fas fa-clipboard-check"
              style="font-size: 2rem; margin-bottom: 1rem;"
            />
            <h6
              style="font-weight: 600; margin-bottom: 0.5rem;"
            >
              Requirements
            </h6>
            <small>
              Step 3
            </small>
          </div>
        </div>
        <div
          class="col-md-6 col-lg-3 mb-3"
        >
          <div
            style="background: white; color: rgb(51, 51, 51); border-radius: 15px; padding: 1.5rem; cursor: pointer; transition: all 0.3s ease; box-shadow: 0 5px 20px rgba(0,0,0,0.1); border: 2px solid #e8f5e9; text-align: center;"
          >
            <i
              class="fas fa-trophy"
              style="font-size: 2rem; margin-bottom: 1rem;"
            />
            <h6
              style="font-weight: 600; margin-bottom: 0.5rem;"
            >
              Achievements
            </h6>
            <small>
              Step 4
            </small>
          </div>
        </div>
      </div>
      <div
        style="width: 100%; height: 6px; background: rgb(224, 224, 224); border-radius: 3px; margin-bottom: 2rem; overflow: hidden;"
      >
        <div
          style="height: 100%; border-radius: 3px; transition: width 0.5s ease; width: 25%;"
        />
      </div>
      <div
        style="display: flex; justify-content: center; gap: 1rem; margin-bottom: 3rem;"
      >
        <div
          style="width: 15px; height: 15px; border-radius: 50%; cursor: pointer; transition: all 0.3s ease; transform: scale(1.3);"
        />
        <div
          style="width: 15px; height: 15px; border-radius: 50%; background: rgb(224, 224, 224); cursor: pointer; transition: all 0.3s ease;"
        />
        <div
          style="width: 15px; height: 15px; border-radius: 50%; background: rgb(224, 224, 224); cursor: pointer; transition: all 0.3s ease;"
        />
        <div
          style="width: 15px; height: 15px; border-radius: 50%; background: rgb(224, 224, 224); cursor: pointer; transition: all 0.3s ease;"
        />
      </div>
      <div
        class="row justify-content-center"
      >
        <div
          class="col-lg-10"
        >
          <div
            class="active-card"
            style="background: white; border-radius: 20px; padding: 2rem; margin: 1rem 0px; box-shadow: 0 20px 60px rgba(76, 175, 80, 0.3); border: 3px solid transparent; transition: all 0.4s ease; cursor: pointer; position: relative; overflow: hidden; border-color: #4caf50; transform: translateY(-10px);"
          >
            <i
              class="fas fa-bullseye"
              style="position: absolute; top: 20px; right: 20px; font-size: 2rem; color: rgb(76, 175, 80); transition: all 0.3s ease; transform: scale(1.2);"
            />
            <div
              style="width: 60px; height: 60px; border-radius: 50%; color: white; display: flex; align-items: center; justify-content: center; font-size: 1.5rem; font-weight: bold; margin: 0px auto 1rem; position: relative; z-index: 2;"
            >
              1
            </div>
            <h3
              style="color: rgb(46, 125, 50); font-weight: 700; font-size: 1.5rem; margin-bottom: 1rem; text-align: center;"
            >
              Aim
            </h3>
            <div
              style="color: rgb(51, 51, 51); line-height: 1.7; font-size: 1rem;"
            >
              <p
                class="mb-3"
                style="text-align: justify;"
              >
                The volunteer program at Hope Harvest aims to create a network of dedicated individuals who contribute their time, skills, and passion to make a tangible difference in communities across Bangladesh. We believe in the power of collective action and that every person has something valuable to offer in service to others. Our goal is to harness this potential by providing structured opportunities for meaningful engagement.
              </p>
              <p
                class="mb-3"
                style="text-align: justify;"
              >
                We strive to build a diverse volunteer base that represents all segments of society, including students, professionals, retirees, and homemakers. By bringing together people from different backgrounds, we create a rich pool of skills, perspectives, and experiences that enhance our ability to address complex community challenges. Our volunteer program is designed to be inclusive, welcoming anyone with a sincere desire to help regardless of their previous experience.
              </p>
              <p
                class="mb-3"
                style="text-align: justify;"
              >
                Beyond the immediate impact on beneficiaries, we aim to foster a culture of volunteerism and civic responsibility in Bangladesh. Through volunteering with Hope Harvest, participants develop empathy, leadership skills, and a deeper understanding of social issues. Many volunteers report that their experiences transform their worldview and inspire continued community involvement throughout their lives.
              </p>
              <p
                class="mb-3"
                style="text-align: justify;"
              >
                The program also serves as a bridge between different communities, creating connections between those who serve and those who are served. These relationships help break down stereotypes, build mutual respect, and create a stronger sense of shared humanity. Ultimately, we envision a society where volunteerism is a natural and valued part of citizenship, contributing to social cohesion and collective wellbeing.
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row mt-5"
      >
        <div
          class="col-12 text-center"
        >
          <button
            style="border-radius: 50px; padding: 20px 50px; color: white; font-weight: 700; font-size: 1.3rem; cursor: pointer; transition: all 0.3s ease; box-shadow: 0 8px 30px rgba(76, 175, 80, 0.4); text-transform: uppercase; letter-spacing: 2px;"
          >
            <i
              class="fas fa-hands-helping me-2"
            />
            Join as Volunteer
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
